### PR TITLE
Replace panic in `datafusion-expr` crate

### LINKED ISSUE #3312: Review use of panic in `datafusion-expr` crate
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

```
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/binary_rule.rs":385                 _ => unreachable!(),
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/binary_rule.rs":388         _ => unreachable!(),
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/expr_rewriter.rs":356                     let normalized_expr = normalized_expr.unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/builder.rs":610                 && can_hash(self.plan.schema().field_from_column(l).unwrap().data_type())
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/builder.rs":615                 && can_hash(self.plan.schema().field_from_column(r).unwrap().data_type())
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/builder.rs":632             join.filter(filters.unwrap())
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/builder.rs":912                 .unwrap(),
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/display.rs":234         self.parent_ids.pop().unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/plan.rs":546                 self.0.accept(&mut visitor).unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/plan.rs":587                 self.0.accept(&mut visitor).unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/plan.rs":638                 self.0.accept(&mut visitor).unwrap();
"/home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/plan.rs":643                 self.0.accept(&mut visitor).unwrap();
```

**Describe the solution you'd like**
Review code that can panic and see where it makes sense to return a `Result` instead. For example, It is generally better to use `?` than `unwrap`on results.

The goal is not to remove all panics but review and make sure we are using them appropriately. Bonus points for adding documentation for invariants.

**Describe alternatives you've considered**
None

**Additional context**
File list generated by https://github.com/andygrove/no-need-to-panic

### PR DESCRIPTION
**Pull request by iajoiner**
# Which issue does this PR close?

Closes #3312.

 # Rationale for this change

Replace `panic`s with mostly `Result`s.
# What changes are included in this PR?

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by iajoiner** [comment]
@andygrove Please review. Note that the `unreachable!`s have been replaced with `None` following what other private coercing functions do.

**Comment by iajoiner** [comment]
The Mac test probably needs to be rerun.

**Comment by alamb** [comment]
I restarted the mac test

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -605,14 +605,12 @@ impl LogicalPlanBuilder {
         let mut join_on: Vec<(Column, Column)> = vec![];
         let mut filters: Option<Expr> = None;
         for (l, r) in &on {
-            if self.plan.schema().field_from_column(l).is_ok()
-                && right.schema().field_from_column(r).is_ok()
-                && can_hash(self.plan.schema().field_from_column(l).unwrap().data_type())
+            if right.schema().field_from_column(r).is_ok()
```
Comment (andygrove) [inline datafusion/expr/src/logical_plan/builder.rs:608]: This looks like it has changed the semantics of the code. Previously if `field_from_column(l)` was an `Err` then a different code path would have been followed and with this change, there would be an error instead.

Comment (iajoiner) [reply datafusion/expr/src/logical_plan/builder.rs:608]: Ah I see. Really thanks! Fixed!

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @iajoiner

**Pull request merged**